core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -155,90 +155,19 @@
   private final HashSet<DelayedClientTransport> delayedTransports =
       new HashSet<DelayedClientTransport>();
 
-  @VisibleForTesting
-  final InUseStateAggregator<Object> inUseStateAggregator =
-      new InUseStateAggregator<Object>() {
-        @Override
-        Object getLock() {
-          return lock;
-        }
+  private final Future<?> idleModeTimerFuture;
 
-        @Override
-        @GuardedBy("lock")
-        void handleInUse() {
-          exitIdleMode();
-        }
+  // Set every time a load balancer gets a transport.
+  private volatile boolean isIdle;
 
-        @GuardedBy("lock")
-        @Override
-        void handleNotInUse() {
-          if (shutdown) {
-            return;
-          }
-          rescheduleIdleTimer();
-        }
-      };
-
-  private class IdleModeTimer implements Runnable {
-    @GuardedBy("lock")
-    boolean cancelled;
-
-    @Override
-    public void run() {
-      ArrayList<TransportSet> transportsCopy = new ArrayList<TransportSet>();
-      LoadBalancer<ClientTransport> savedBalancer;
-      NameResolver oldResolver;
-      synchronized (lock) {
-        if (cancelled) {
-          // Race detected: this task started before cancelIdleTimer() could cancel it.
-          return;
-        }
-        // Enter idle mode
-        savedBalancer = loadBalancer;
-        loadBalancer = null;
-        oldResolver = nameResolver;
-        nameResolver = getNameResolver(target, nameResolverFactory, nameResolverParams);
-        transportsCopy.addAll(transports.values());
-        transports.clear();
-        decommissionedTransports.addAll(transportsCopy);
-      }
-      for (TransportSet ts : transportsCopy) {
-        ts.shutdown();
-      }
-      savedBalancer.shutdown();
-      oldResolver.shutdown();
+  LoadBalancer<ClientTransport> getLoadBalancer() {
+    isIdle = false;
+    if (loadBalancer != null) {
```

Comment:
loadBalancer could have been cleared before it's returned. This can be easily addressed by copying it to a local variable before checking it.

However, there is another race where another thread is in `GoIdle` and shuts down loadBalancer while it being returned, which would result in an spurious error seen by the application.  This is the trickiest part.